### PR TITLE
triple flowcontrol

### PR DESCRIPTION
**Pull request by asa3311**
triple flowcontrol sub 3.2 branch

## Checklist
- [x] Make sure there is a [GitHub_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GitHub issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Check if is necessary to patch to Dubbo 3 if you are work on Dubbo 2.7
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [ ] Add some description to [dubbo-website](https://github.com/apache/dubbo-website) project if you are requesting to add a feature.
- [ ] GitHub Actions works fine on your own branch.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

**Review by guohao — CHANGES_REQUESTED** [review]

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/TriHttp2LocalFlowController.java [hunk 1]
```diff
@@ -0,0 +1,644 @@
...
+    public TriHttp2LocalFlowController(Http2Connection connection,
+                                       float windowUpdateRatio,
+                                       boolean autoRefillConnectionWindow) {
+        this.connection = checkNotNull(connection, "connection");
+        windowUpdateRatio(windowUpdateRatio);
+
+        // Add a flow state for the connection.
+        stateKey = connection.newKey();
+        org.apache.dubbo.rpc.protocol.tri.TriHttp2LocalFlowController.FlowState connectionState = autoRefillConnectionWindow ?
+            new org.apache.dubbo.rpc.protocol.tri.TriHttp2LocalFlowController.AutoRefillState(connection.connectionStream(), initialWindowSize) :
```
[the remark this answers is not in the document]
Comment (asa3311) [reply dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/TriHttp2LocalFlowController.java]: > Why not enable AutoRefillState by default ?

this code from netty

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/TriHttp2LocalFlowController.java [hunk 2]
```diff
@@ -0,0 +1,644 @@
...
+    public void incrementWindowSize(Http2Stream stream, int delta) throws Http2Exception {
+        assert ctx != null && ctx.executor().inEventLoop();
+        org.apache.dubbo.rpc.protocol.tri.TriHttp2LocalFlowController.FlowState state = state(stream);
+        // Just add the delta to the stream-specific initial window size so that the next time the window
+        // expands it will grow to the new initial size.
+        state.incrementInitialStreamWindow(delta);
+        state.writeWindowUpdateIfNeeded();
+    }
+
+    public boolean consumeTriBytes(Http2Stream stream, int numBytes)  throws Http2Exception {
```
[the remark this answers is not in the document]
Comment (asa3311) [reply dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/TriHttp2LocalFlowController.java]: change to void

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/TriHttp2LocalFlowController.java [hunk 3]
```diff
@@ -0,0 +1,644 @@
...
+        @Override
+        public void incrementFlowControlWindows(int delta) throws Http2Exception {
+            if (delta > 0 && window > MAX_INITIAL_WINDOW_SIZE - delta) {
+                throw streamError(stream.id(), FLOW_CONTROL_ERROR,
+                    "Flow control window overflowed for stream: %d", stream.id());
+            }
+
+            window += delta;
+            processedWindow += delta;
+            lowerBound = delta < 0 ? delta : 0;
```
[the remark this answers is not in the document]
Comment (asa3311) [reply dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/TriHttp2LocalFlowController.java]: done

**Review by guohao — CHANGES_REQUESTED** [review]
Maybe a performance comparison report is needed for introducing custom flow control.

File: dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/beans/factory/annotation/ServiceAnnotationPostProcessor.java [hunk 4]
```diff
@@ -354,7 +355,12 @@ private void processScannedBeanDefinition(BeanDefinitionHolder beanDefinitionHol
     private Annotation findServiceAnnotation(Class<?> beanClass) {
         return serviceAnnotationTypes
                 .stream()
-                .map(annotationType -> findAnnotation(beanClass, annotationType))
+            .map(annotationType ->
```
[the remark this answers is not in the document]
Comment (asa3311) [reply dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/beans/factory/annotation/ServiceAnnotationPostProcessor.java]: Removed

File: dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcServiceContext.java [hunk 5]
```diff
@@ -186,6 +186,9 @@ public List<URL> getUrls() {
     @Override
     public void setUrls(List<URL> urls) {
         this.urls = urls;
+        if (!urls.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (asa3311) [reply dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcServiceContext.java]: Removed

File: dubbo-rpc/dubbo-rpc-dubbo/src/main/java/org/apache/dubbo/rpc/protocol/dubbo/DubboInvoker.java [hunk 6]
```diff
@@ -108,7 +115,7 @@ protected Result doInvoke(final Invocation invocation) throws Throwable {
             } else {
                 ExecutorService executor = getCallbackExecutor(getUrl(), inv);
                 CompletableFuture<AppResponse> appResponseFuture =
-                        currentClient.request(inv, timeout, executor).thenApply(obj -> (AppResponse) obj);
+                    currentClient.request(inv, timeout, executor).thenApply(obj -> (AppResponse) obj);
```
[the remark this answers is not in the document]
Comment (asa3311) [reply dubbo-rpc/dubbo-rpc-dubbo/src/main/java/org/apache/dubbo/rpc/protocol/dubbo/DubboInvoker.java]: Removed

File: dubbo-rpc/dubbo-rpc-injvm/src/main/java/org/apache/dubbo/rpc/protocol/injvm/InjvmInvoker.java [hunk 7]
```diff
@@ -102,7 +102,15 @@ public Result doInvoke(Invocation invocation) throws Throwable {
         if (serverHasToken) {
             invocation.setAttachment(Constants.TOKEN_KEY, serverURL.getParameter(Constants.TOKEN_KEY));
         }
-        invocation.setAttachment(TIMEOUT_KEY, calculateTimeout(invocation, invocation.getMethodName()));
+
+        int timeout = calculateTimeout(invocation, invocation.getMethodName());
```
[the remark this answers is not in the document]
Comment (asa3311) [reply dubbo-rpc/dubbo-rpc-injvm/src/main/java/org/apache/dubbo/rpc/protocol/injvm/InjvmInvoker.java]: Removed

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/TriHttp2LocalFlowController.java [hunk 8]
```diff
@@ -0,0 +1,648 @@
...
+                // Unconditionally used the reduced flow control state because it requires no object allocation
+                // and the DefaultFlowState will be allocated in onStreamActive.
+                stream.setProperty(stateKey, REDUCED_FLOW_STATE);
+            }
+
+            @Override
+            public void onStreamActive(Http2Stream stream) {
+                // Need to be sure the stream's initial window is adjusted for SETTINGS
+                // frames which may have been exchanged while it was in IDLE
+                stream.setProperty(stateKey, new org.apache.dubbo.rpc.protocol.tri.TriHttp2LocalFlowController.DefaultState(stream, initialWindowSize));
```
[the remark this answers is not in the document]
Comment (asa3311) [reply dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/TriHttp2LocalFlowController.java]: done

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/TripleInvoker.java [hunk 9]
```diff
@@ -192,8 +192,15 @@ StreamObserver<Object> streamCall(ClientCall call,
     AsyncRpcResult invokeUnary(MethodDescriptor methodDescriptor, Invocation invocation,
         ClientCall call) {
         ExecutorService callbackExecutor = getCallbackExecutor(getUrl(), invocation);
+
         int timeout = calculateTimeout(invocation, invocation.getMethodName());
+        if (timeout <= 0) {
```
[the remark this answers is not in the document]
Comment (asa3311) [reply dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/TripleInvoker.java]: Removed

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/TripleProtocol.java [hunk 10]
```diff
@@ -48,6 +48,7 @@
 import java.util.Set;
 import java.util.concurrent.ExecutorService;
 
+
```
[the remark this answers is not in the document]
Comment (asa3311) [reply dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/TripleProtocol.java]: done

**Pull request merged**